### ISSUE TITLE
Possible missing AuditStamp field in CreateDevice()

### ISSUE THREAD

**Issue by AndHam89** [samilliken/openDCIM#457]
Is it possible that the new field 'AuditStamp' is not fully managed in class Device? Trying to add a device in 3.2.1 I get the following apache error.

```log
Mon May 05 22:00:35.636424 2014] [php5:error] [pid 22536] [client 192.168.1.2:50667] PDO Error: Field 'AuditStamp' doesn't have a default value SQL=INSERT INTO fac_Device SET Label="SRV-TEST-AF01", SerialNo="", AssetTag="", \r\n\t\t\tPrimaryIP="", SNMPCommunity="", ESX=0, Owner=0, \r\n\t\t\tEscalationTimeID=0, EscalationID=0, PrimaryContact=0, \r\n\t\t\tCabinet=2947, Position=4, Height=2, Ports=10, \r\n\t\t\tFirstPortNum=0, TemplateID=1337, NominalWatts=500, \r\n\t\t\tPowerSupplyCount=2, DeviceType="Server", ChassisSlots=0, \r\n\t\t\tRearChassisSlots=0,ParentDevice=0, \r\n\t\t\tMfgDate="1970-01-01", \r\n\t\t\tInstallDate="2014-05-05", WarrantyCo="", \r\n\t\t\tWarrantyExpire="1970-01-01", Notes="", \r\n\t\t\tReservation=0, HalfDepth=0, BackSide=0;, referer: https://dcim321.site/devices.php?action=new&cabinet=2947
```

Could it be that something like the following needs to be added?

```diff
--- assets.inc.php.orig 2014-04-14 22:27:50.000000000 +0200
+++ assets.inc.php      2014-05-05 22:24:55.235172727 +0200
@@ -1111,6 +1111,7 @@
                $this->Reservation=intval($this->Reservation);
                $this->HalfDepth=intval($this->HalfDepth);
                $this->BackSide=intval($this->BackSide);
+                $this->AuditStamp=sanitize($this->AuditStamp);
        }

        function MakeDisplay() {
@@ -1121,7 +1122,7 @@
                $this->SNMPCommunity=stripslashes($this->SNMPCommunity);
                $this->MfgDate=stripslashes($this->MfgDate);
                $this->InstallDate=stripslashes($this->InstallDate);
-               $this->WarrantyCo=stripslashes($this->WarrantyCo);
+                $this->WarrantyCo=stripslashes($this->WarrantyCo);
                $this->WarrantyExpire=stripslashes($this->WarrantyExpire);
                $this->Notes=stripslashes($this->Notes);
        }
@@ -1236,7 +1237,7 @@
                        MfgDate=\"".date("Y-m-d", strtotime($this->MfgDate))."\",
                        InstallDate=\"".date("Y-m-d", strtotime($this->InstallDate))."\", WarrantyCo=\"$this->WarrantyCo\",
                        WarrantyExpire=\"".date("Y-m-d", strtotime($this->WarrantyExpire))."\", Notes=\"$this->Notes\",
-                       Reservation=$this->Reservation, HalfDepth=$this->HalfDepth, BackSide=$this->BackSide;";
+                       Reservation=$this->Reservation, HalfDepth=$this->HalfDepth, BackSide=$this->BackSide, AuditStamp=now();";

                if ( ! $dbh->exec( $sql ) ) {
                        $info = $dbh->errorInfo();
```

**Comment by TafThorne** [comment]
I am seeing something very like this on dcim/4.3 since I upgraded the server from Ubuntu 14.04 to 16.04.  When I try to Add Device things go OK until I hit the Create button, where it silently fails (on the web interface).  When I check the error log I can see:

```
[Fri Feb 24 11:19:32.639025 2017] [:error] [pid 32267] [client 172.16.20.146:53930] PHP Notice:  Use of undefined constant MB_CASE_UPPER - assumed 'MB_CASE_UPPER' in /var/www/openDCIM-4.3/misc.inc.php on line 259, referer: https://<opendcim.url>/devices.php?action=new&CabinetID=2
[Fri Feb 24 11:19:32.639639 2017] [:error] [pid 32267] [client 172.16.20.146:53930] PHP Notice:  Use of undefined constant MB_CASE_UPPER - assumed 'MB_CASE_UPPER' in /var/www/openDCIM-4.3/misc.inc.php on line 259, referer: https://<opendcim.url>/devices.php?action=new&CabinetID=2
[Fri Feb 24 11:19:32.640165 2017] [:error] [pid 32267] [client 172.16.20.146:53930] PHP Notice:  Use of undefined constant MB_CASE_UPPER - assumed 'MB_CASE_UPPER' in /var/www/openDCIM-4.3/misc.inc.php on line 259, referer: https://<opendcim.url>/devices.php?action=new&CabinetID=2
[Fri Feb 24 11:19:32.640508 2017] [:error] [pid 32267] [client 172.16.20.146:53930] PDO Error: Field 'AuditStamp' doesn't have a default value SQL=INSERT INTO fac_Device SET Label="SWITCH09",  \n\t\t\tAssetTag="", PrimaryIP="172.16.20.9", \n\t\t\tSNMPCommunity="", SNMPVersion="2c",\n\t\t\tv3SecurityLevel="noAuthNoPriv", EscalationID=0, \n\t\t\tv3AuthProtocol="MD5", Position=25, \n\t\t\tv3AuthPassphrase="", DeviceType="Switch",\n\t\t\tv3PrivProtocol="DES", NominalWatts=482, \n\t\t\tv3PrivPassphrase="", Weight=5,\n\t\t\tSNMPFailureCount=0, ESX=0, Owner=1, \n\t\t\tEscalationTimeID=0, PrimaryContact=0, \n\t\t\tCabinet=2, Height=1, Ports=50, \n\t\t\tFirstPortNum=0, TemplateID=5, \n\t\t\tPowerSupplyCount=1, ChassisSlots=0, \n\t\t\tRearChassisSlots=0,ParentDevice=0,\n\t\t\tMfgDate="1970-01-01", \n\t\t\tInstallDate="2016-07-21", \n\t\t\tWarrantyCo="", Notes="", \n\t\t\tWarrantyExpire="1970-01-01", \n\t\t\tReservation=0, HalfDepth=0, \n\t\t\tBackSide=0, SerialNo="";, referer: https://<opendcim.url>/devices.php?action=new&CabinetID=2
```

With `PDO Error: Field 'AuditStamp' doesn't have a default value SQL=INSERT INTO fac_Device SET Label="SWITCH09"` being the similar looking error.

**Comment by TafThorne** [comment]
Ah.  I have tried upgrading to dcim/4.3.1 and as part of that I have now seen the warning 
```
openDCIM does not support STRICT_TRANS_TABLES. The following SQL statement might clear the error for this session. More information can be found here.

SET GLOBAL sql_mode = "";
``` 
Which also links back to this issue.  I shall attempt to follow the work around.

**Comment by TafThorne** [comment]
On my Debian 16.04 machine, in `/etc/mysql/conf.d/mysql.cnf` I added the lines
```
[mysqld]
sql_mode=
```

Stopped and started the mysql service, then restarted the apache2 service and the error report went away. 

After completing my install / update I could then successfully add devices again.

**Comment by valffadir** [comment]
I hope this can help those who still got this problem:

I got the same message when I tried to install openDCIM:

`openDCIM does not support STRICT_TRANS_TABLES. The following SQL statement might clear the error for this session. More information can be found here.`

I don't have privileges on the server I had to install opendcim on, so I couldn't change sql_mode with 
`SET GLOBAL sql_mode = "";` either. The only solution I coud find is to set the SESSION variable `sql_mode` to `""` because you don't need privileges to do it; but changing a SESSION variable means that it will return to it's original value (the GLOBAL value) every time you close the connection with mysql.

I used this trick to solve it: 

- I modified the `db.inc.php` file adding:

`$dbh->exec("SET @@SESSION.sql_mode = ''");`

under `$dbh=new PDO($pdoconnect,$dbuser,$dbpass,$pdo_options);`. 
This sets the value of your SESSION variable `sql_mode` to `""` every time you establish the connection with the DB.

- Then, I modified the `preflight.inc.php` file changing:
`// check for strict_trans_tables`
`if(strpos(@end($dbh->query("select @@global.sql_mode;")->fetch()),'STRICT_TRANS_TABLES') === false){
`
to
`// check for strict_trans_tables`
`if(strpos(@end($dbh->query("select @@session.sql_mode;")->fetch()),'STRICT_TRANS_TABLES') === false){
`
just to have that pretty table filled with GOOD values.

I hope the devs could think consider adding this solution to upcoming releases.
